File: archive/piranesi/legacy/tests/pre-pivot/fixtures/owasp/deserialization.py
import marshal
import pickle

import yaml
from flask import Flask, request

app = Flask(__name__)


@app.route("/api/load", methods=["POST"])
def load_pickle():
    data = request.get_data()  # tainted
    obj = pickle.loads(data)  # CWE-502 sink - critical
    return str(obj)


@app.route("/api/yaml", methods=["POST"])
def load_yaml():
    raw = request.get_data(as_text=True)  # tainted
    parsed = yaml.load(raw, Loader=yaml.Loader)  # CWE-502 sink - unsafe loader
    return str(parsed)


@app.route("/api/marshal", methods=["POST"])
def load_marshal():
    data = request.get_data()  # tainted
    obj = marshal.loads(data)  # CWE-502 sink
    return str(obj)


# SAFE: yaml.safe_load
@app.route("/api/safe-yaml", methods=["POST"])
def safe_load_yaml():
    raw = request.get_data(as_text=True)
    parsed = yaml.safe_load(raw)  # safe loader - sanitizer
    return str(parsed)


# SAFE: pickle with HMAC verification (not modeled, but no taint from user)
@app.route("/api/safe-pickle")
def safe_pickle():
    with open("trusted_data.pkl", "rb") as f:
        obj = pickle.load(f)  # not from user input
    return str(obj)

File: archive/piranesi/legacy/tests/pre-pivot/fixtures/owasp/deserialization.ts
import express from "express";
import yaml from "js-yaml";

const app = express();

// VULNERABLE: JSON.parse on raw user input without schema validation
app.post("/api/config", (req, res) => {
  const rawBody = req.body.config; // tainted
  const parsed = JSON.parse(rawBody); // CWE-502 sink
  applyConfig(parsed);
  res.json({ ok: true });
});

// VULNERABLE: yaml.load on user input
app.post("/api/import", (req, res) => {
  const yamlData = req.body.data; // tainted
  const parsed = yaml.load(yamlData); // CWE-502 sink (unsafe loader)
  res.json(parsed);
});

// SAFE: JSON.parse followed by schema validation (Joi)
app.post("/api/safe-config", (req, res) => {
  const rawBody = req.body.config;
  const parsed = JSON.parse(rawBody);
  const validated = schema.validate(parsed); // sanitizer
  if (validated.error) return res.status(400).json(validated.error);
  applyConfig(validated.value);
  res.json({ ok: true });
});

// SAFE: yaml.safeLoad
app.post("/api/safe-import", (req, res) => {
  const yamlData = req.body.data;
  const parsed = yaml.safeLoad(yamlData); // safe loader
  res.json(parsed);
});

function applyConfig(config: any) {}
const schema = { validate: (d: any) => ({ error: null, value: d }) };
